File: backend/src/deaths_door/character.py
from pydantic import BaseModel, Field

from .alignment import Alignment
from .changes import Changes
from .character_type import CharacterType
from .status_effects import StatusEffect, StatusEffectOut


class CharacterOut(BaseModel):
    """A character with only fields meant to be sent to the client."""

    name: str = Field(
        ...,
        description="Character's display name",
        examples=["Imp", "Chef", "Empath", "Baron"],
    )
    description: str = Field(
        ...,
        description="Character's ability description",
        examples=[
            "Each night*, choose a player: they die. If you kill yourself this way, a Minion becomes the Imp."
        ],
    )
    icon_path: str = Field(
        ...,
        description="Path to the character's icon image",
        examples=["imp.png", "chef.png"],
    )
    alignment: Alignment = Field(
        ...,
        description="Character's alignment (good or evil)",
        examples=["evil", "good"],
    )
    category: CharacterType = Field(
        ...,
        description="Character's type/category",
        examples=["demon", "townsfolk", "outsider", "minion", "traveler"],
    )


class Character:
    """A character in the script."""

    name: str
    description: str
    category: CharacterType
    alignment: Alignment
    status_effects: list[StatusEffect]
    changes: None | Changes

    def __str__(self) -> str:
        """Return a human-readable string representation."""
        return f"{self.name} ({self.category.value})"

    def __repr__(self) -> str:
        """Return a detailed string representation."""
        return (
            f"Character(name={self.name!r}, "
            f"category={self.category!r}, "
            f"alignment={self.alignment!r})"
        )

    def normalize_name_for_comparison(self, name: str) -> str:
        """Convert name to lowercase and remove whitespace for comparison."""
        return name.lower().strip()

    def is_named(self, name: str) -> bool:
        """Check if the character matches the given name."""
        return self.normalize_name_for_comparison(
            self.name
        ) == self.normalize_name_for_comparison(name)

    def get_icon_path(self) -> str:
        """Return the character's icon."""
        return self.name.lower().replace(" ", "") + ".png"

    def get_status_effects_out(self) -> list[StatusEffectOut]:
        """Return the character's status effects."""
        return [
            status_effect.to_out(self.name) for status_effect in self.status_effects
        ]

    def to_out(self) -> CharacterOut:
        """Convert the character to a character out."""
        return CharacterOut(
            name=self.name,
            description=self.description,
            icon_path=self.get_icon_path(),
            alignment=self.alignment,
            category=self.category,
        )

File: backend/src/character.rs
//! A character in a script, plus its API output model.

use serde::{Deserialize, Serialize};

use crate::alignment::Alignment;
use crate::character_type::CharacterType;
use crate::status_effect::StatusEffectOut;

/// A character with only the fields meant to be sent to the client.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, utoipa::ToSchema)]
pub struct CharacterOut {
    pub name: String,
    pub description: String,
    pub icon_path: String,
    pub alignment: Alignment,
    pub category: CharacterType,
}

/// A character in the script. The icon filename is derived on demand via
/// [`Character::get_icon_path`]; any `icon_path`/`changes` keys in the source
/// data are ignored (serde ignores unknown fields by default).
#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct Character {
    pub name: String,
    pub description: String,
    pub category: CharacterType,
    pub alignment: Alignment,
    /// Names of the status effects this character can introduce.
    #[serde(default)]
    pub status_effects: Vec<String>,
}

impl Character {
    fn normalize(name: &str) -> String {
        name.to_lowercase().trim().to_string()
    }

    /// Check if the character matches the given name (case-insensitive, trimmed).
    pub fn is_named(&self, name: &str) -> bool {
        Self::normalize(&self.name) == Self::normalize(name)
    }

    /// The character's icon filename, matching Python's `get_icon_path`.
    pub fn get_icon_path(&self) -> String {
        format!("{}.png", self.name.to_lowercase().replace(' ', ""))
    }

    /// The character's status effects as API output models.
    pub fn get_status_effects_out(&self) -> Vec<StatusEffectOut> {
        self.status_effects
            .iter()
            .map(|effect| StatusEffectOut {
                name: effect.clone(),
                character_name: self.name.clone(),
            })
            .collect()
    }

    /// Convert the character to its API output model.
    pub fn to_out(&self) -> CharacterOut {
        CharacterOut {
            name: self.name.clone(),
            description: self.description.clone(),
            icon_path: self.get_icon_path(),
            alignment: self.alignment,
            category: self.category,
        }
    }
}
